### DISCUSSION TITLE
Backup of face-tagging possible?

### DISCUSSION THREAD

**Discussion by poisonnuke** [immich-app/immich#10037] [General]
I've spent a great deal of time to name and merge all the people (and hide thousands of not relevant people). I would like to create a backup of that data so in case the server goes down or an immich upgrade (or change or ML model fails) I can restore that.

My biggest concern really is that currently a change of the ML-model will just ditch everything one did and that could happen with an upgrade too.

**Comment by bo0tzz** [comment]
The face data is stored in the database, which you should absolutely be backing up. https://immich.app/docs/administration/backup-and-restore

**Reply by poisonnuke** [reply]
the bigger concern for me is: will this backup always work? I read multiple times that changing the ML model behind the tagging will effectively render the old data useless.

Just assuming a new version of immich will bring a new ML model for the face-tagging, what will happen during an upgrade and when performing a fresh install and restoring a backup? 

is there no standard to save such data in the pictures itself?

**Reply by bo0tzz** [reply]
The ML model isn't just randomly going to change, if we do change the default (which I don't see happening anytime soon) it'll clearly be called out in the release notes (which you should always read before upgrading).

> is there no standard to save such data in the pictures itself?

There is no good standardization for this as far as I'm aware, no. See also the discussion in #4348

**Reply by poisonnuke** [reply]
im not really into how ML models work, do you think there will be possibilities to use the data from one model with another model? Maybe someone could develop a something like a migration script that translates the findings of one model to the ones of another?

**Reply by bo0tzz** [reply]
That's not really possible, no. This isn't something you need to be worried about though.

**Reply by rhatguy** [reply]
There is a lot of good work going on in #6455 that might apply here.  If Immich eventually writes this data into the metadata of the picture (or into an XMP file) then the data will become portable between different applications or potentially between different ML models.

There was also a previous feature request under #4348

**Comment by nomandera** [comment]
I want to put a different spin on this.

Face tagging is human driven and humans do not live forever. At some point it will be quite literally impossible for an AI to be retrained with face names and they will be lost to time forever.

I think this adds a unique high value to this data that makes it worth specific consideration for a method of preservation (data Anthropology?) that can stand the test of time, perhaps even generations.

In essence I think we are talking here about exporting training sets.

**Reply by rhatguy** [reply]
I think the right long term answer to this is to write the face metadata into the files the pictures are stored in (or in sidecar files that live alongside the files).  I view this as the digital equivalent of writing on a printed picture.  Storing the metadata along with the photo/file has been a generally industry accepted methodology for some time.  It started with storing the creation date, evolved into storing the (GPS) location of the photo and seems to be evolving again into storing face tags as a lot of software are starting to interoperate with face metadata.

The methods of generating this metadata have evolved over time.  I can remember my first digital camera which didn't store the creation date of the photo.  These days its almost crazy to think of a camera not automatically tagging the date/time a photo was created.  Similar things happened with location metadata.  First people added it to photos after the fact.  Generally those capabilities started to be added to the devices that took those photos, until today we expect our phones to automatically tag the location without really thinking about it.  Face metadata seems to be going through a similar transformation where initially it was added by humans without machine assistance.  These days there are many accepted methodologies of creating face metadata with AI playing a large role in its creation.

I believe eventually we will look at face metadata just as we do creation date today.  Its an essential piece of metadata that will be stored with our photos and read ubiquitously by all software that interacts with the photos.  Immich has taken the great first step of reading that metadata.  Now there are a host of tools that can create that metadata that Immich can consume.  Hopefully eventually Immich will allow users to store metadata that Immich created (through AI/ML) in that standardized format inside the pictures.  Maybe eventually even the large players like Apple Photos and Google photos will adopt similarly portable metadata formats!

**Reply by nomandera** [reply]
I took some time to ponder this before I jumped to reply and on the whole after some consideration I agree.

I personally would only opt for sidecar files as I strongly believe the original photo should remain a bit perfect copy from the imaging device with exact timestamps wherever technically possible.

Obviously we run the risk with whatever approach we opt for if the data storage type becoming old fashioned. For instance I spent a lot of time in Picassa tagging only to lose most of it due to lack of elegant way to recover. In the end this didnt matter as the state of the art increased so significantly I could train an AI to do better than me for most of it quickly but it describes a very real risk.

Sidecar files even today are an absolute mess of standards but they likely still offer the best hope.

I see many mentions of complete sidecar support in Immich and real features starting to emerge so I can hope this continues to a viable solution for me that would have three key features:

1. Store all tagging in the sidecar and update it real time.
2. Allow sidecar files to live somewhere other than beside the image to allow for RO images, storing sidecar on fast IO SSD and regular snapshots
3. Add a CRC and name of the image to the sidecar on its creation to allow for definitive relation to the specific image (duplicate names are common here with family phones) and a means to CRC check the integrity of the image at any time to check for bitrot or manipulation.

For now I think I need to ponder what to do until this support is added. I already have faces tagged by family that I could not replicate myself. In fact some are class photos from my kids and even now I bet they couldn't identify them all.

**Reply by rhatguy** [reply]
You're mention of Picasa demonstrates the exact point of keeping metadata with or in the files.  Picasa had a feature to write the face metadata into the images.  If that feature were enabled, those same face tags are still visible today in Immich (now that it supports reading that metadata).  I personally have pictures that I tagged in Picasa many years on (on Windows XP :) ) that I have imported into Immich now.  When Picasa was EOL, I moved my tag creation process to Digikam.  Recently I wanted to switch to a software that worked on mobile (Digikam does not) so I switched to Mylio.  I also use PiGallery2 which also reads that same face metadata.  If you are fully committed to "metadata with/in the image", it does limit your choice of software to use, but that metadata can live forever along with the picture itself.  The nice thing about being committed to keeping the metadata in/with the picture, is you can change both the way the metadata is created (AI or manual) or the way it's displayed as software comes and goes and you don't loose any metadata you've curated.
